File: blue-core/src/blue/SoundLayer.java
/*
 * blue - object composition environment for csound Copyright (c) 2001-2003
 * Steven Yi (stevenyi@gmail.com)
 *
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License or (at your option) any
 * later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.LIB. If not, write to the Free
 * Software Foundation Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307
 * USA
 */

package blue;

/**
 * Title: blue (Object Composition Environment) Description: Copyright:
 * Copyright (c) steven yi Company: steven yi music
 * 
 * @author steven yi
 * @version 1.0
 */

import blue.automation.ParameterIdList;
import blue.noteProcessor.NoteProcessorChain;
import blue.noteProcessor.NoteProcessorException;
import blue.score.layers.Layer;
import blue.soundObject.NoteList;
import blue.soundObject.SoundObject;
import blue.utility.ObjectUtilities;
import blue.utility.ScoreUtilities;
import electric.xml.Element;
import electric.xml.Elements;
import java.beans.PropertyChangeEvent;
import java.beans.PropertyChangeListener;
import java.util.*;

public final class SoundLayer implements java.io.Serializable, Layer {

    private transient Vector<PropertyChangeListener> propListeners = null;

    private transient Vector<SoundLayerListener> layerListeners = null;

    private static Comparator sObjComparator = new Comparator() {

        public int compare(Object arg0, Object arg1) {
            SoundObject a = (SoundObject) arg0;
            SoundObject b = (SoundObject) arg1;

            float aStart = a.getStartTime();
            float bStart = b.getStartTime();

            if (aStart > bStart) {
                return 1;
            } else if (aStart < bStart) {
                return -1;
            }

            return 0;

        }

    };

    private ArrayList<SoundObject> soundObjects = new ArrayList<>();

    private ParameterIdList automationParameters = new ParameterIdList();

    private String name;

    private boolean muted;

    private boolean solo;

    private NoteProcessorChain npc = new NoteProcessorChain();

    private int heightIndex = 0;

    public static int HEIGHT_MAX_INDEX = 9;

    // constructor
    public SoundLayer() {
        muted = false;
        solo = false;
        name = "";
    }

    public boolean isMuted() {
        return muted;
    }

    public void setMuted(boolean muted) {
        this.muted = muted;
    }

    public boolean isSolo() {
        return solo;
    }

    public void setSolo(boolean solo) {
        this.solo = solo;
    }

    public void setName(String name) {
        String oldName = this.name;
        this.name = name;
        
        if(!this.name.equals(oldName)) {
            firePropertyChangeEvent(new PropertyChangeEvent(this, "name",
                    oldName, name));
        }
    }

    public String getName() {
        return name;
    }

    public ParameterIdList getAutomationParameters() {
        return automationParameters;
    }

    public void addSoundObject(SoundObject sObj) {
        soundObjects.add(sObj);
        fireSoundObjectAdded(sObj);
    }

    public void removeSoundObject(SoundObject sObj) {
        soundObjects.remove((soundObjects.indexOf(sObj)));
        fireSoundObjectRemoved(sObj);
    }

    public ArrayList<SoundObject> getSoundObjects() {
        return soundObjects;
    }

    public boolean contains(SoundObject sObj) {
        return soundObjects.contains(sObj);
    }

    /**
     * called by PolyObject::getMaxTime()
     */

    public float getMaxTime() {
        return ScoreUtilities.getMaxTime(soundObjects);
    }

    public static SoundLayer loadFromXML(Element data,
            Map<String, Object> objRefMap) throws Exception {
        SoundLayer sLayer = new SoundLayer();

        sLayer.setName(data.getAttributeValue("name"));
        sLayer.setMuted(Boolean.valueOf(data.getAttributeValue("muted"))
                .booleanValue());
        sLayer.setSolo(Boolean.valueOf(data.getAttributeValue("solo"))
                .booleanValue());

        String heightIndexStr = data.getAttributeValue("heightIndex");
        if (heightIndexStr != null) {
            sLayer.setHeightIndex(Integer.parseInt(heightIndexStr));
        }

        Element npcNode = data.getElement("noteProcessorChain");

        if (npcNode != null) {
            sLayer.setNoteProcessorChain(NoteProcessorChain
                    .loadFromXML(npcNode));
        }

        Elements sObjects = data.getElements("soundObject");

        while (sObjects.hasMoreElements()) {
            Object obj = ObjectUtilities.loadFromXML(sObjects.next(),
                    objRefMap);
            sLayer.addSoundObject((SoundObject) obj);
        }

        Elements parameters = data.getElements("parameterId");

        while (parameters.hasMoreElements()) {
            String id = parameters.next().getTextString();
            sLayer.automationParameters.addParameterId(id);
        }

        return sLayer;
    }

    public Element saveAsXML(Map<Object, String> objRefMap) {
        Element retVal = new Element("soundLayer");
        retVal.setAttribute("name", this.getName());
        retVal.setAttribute("muted", Boolean.toString(this.isMuted()));
        retVal.setAttribute("solo", Boolean.toString(this.isSolo()));
        retVal.setAttribute("heightIndex", Integer.toString(this
                .getHeightIndex()));

        retVal.addElement(npc.saveAsXML());

        for (Iterator iter = soundObjects.iterator(); iter.hasNext();) {
            SoundObject sObj = (SoundObject) iter.next();
            retVal.addElement(sObj.saveAsXML(objRefMap));
        }

        for (Iterator iter = automationParameters.iterator(); iter.hasNext();) {
            String id = (String) iter.next();
            retVal.addElement("parameterId").setText(id);
        }

        return retVal;
    }

    @Override
    public Object clone() {
        SoundLayer sLayer = new SoundLayer();

        sLayer.setMuted(this.isMuted());
        sLayer.setSolo(this.isSolo());
        sLayer.setName(this.getName());
        sLayer.setNoteProcessorChain((NoteProcessorChain) this.npc.clone());

        for (Iterator iter = this.soundObjects.iterator(); iter.hasNext();) {
            SoundObject sObj = (SoundObject) iter.next();
            sLayer.addSoundObject((SoundObject) sObj.clone());
        }

        for (Iterator iter = this.automationParameters.iterator(); iter
                .hasNext();) {
            String parameterId = (String) iter.next();
            sLayer.automationParameters.addParameterId(parameterId);
        }

        return sLayer;
    }

    public NoteProcessorChain getNoteProcessorChain() {
        return npc;
    }

    public void setNoteProcessorChain(NoteProcessorChain npc) {
        this.npc = npc;
    }

    public int getHeightIndex() {
        return heightIndex;
    }

    public void setHeightIndex(int heightIndex) {
        if (this.heightIndex == heightIndex) {
            return;
        }

        int oldHeight = this.heightIndex;
        this.heightIndex = heightIndex;

        PropertyChangeEvent pce = new PropertyChangeEvent(this, "heightIndex",
                new Integer(oldHeight), new Integer(heightIndex));

        firePropertyChangeEvent(pce);
    }

    public int getSoundLayerHeight() {
        return (heightIndex + 1) * LAYER_HEIGHT;
    }

    /* Property Change Event Code */

    private void firePropertyChangeEvent(PropertyChangeEvent pce) {
        if (propListeners == null) {
            return;
        }

        Iterator iter = new Vector(propListeners).iterator();

        while (iter.hasNext()) {
            PropertyChangeListener listener = (PropertyChangeListener) iter
                    .next();

            listener.propertyChange(pce);
        }
    }

    public void addPropertyChangeListener(PropertyChangeListener pcl) {
        if (propListeners == null) {
            propListeners = new Vector<>();
        }

        if (propListeners.contains(pcl)) {
            return;
        }

        propListeners.add(pcl);
    }

    public void removePropertyChangeListener(PropertyChangeListener pcl) {
        if (propListeners == null) {
            return;
        }
        propListeners.remove(pcl);
    }

    /* Support for SoundLayerListeners, used by automation */

    public void addSoundLayerListener(SoundLayerListener listener) {
        if (layerListeners == null) {
            layerListeners = new Vector<>();
        }

        layerListeners.add(listener);
    }

    public void removeSoundLayerListener(SoundLayerListener listener) {
        if (layerListeners != null) {
            layerListeners.remove(listener);
        }
    }

    private void fireSoundObjectAdded(SoundObject sObj) {
        if (layerListeners != null) {
            Iterator iter = new Vector(layerListeners).iterator();

            while (iter.hasNext()) {
                SoundLayerListener listener = (SoundLayerListener) iter.next();
                listener.soundObjectAdded(this, sObj);
            }
        }
    }

    private void fireSoundObjectRemoved(SoundObject sObj) {
        if (layerListeners != null) {
            Iterator iter = new Vector(layerListeners).iterator();

            while (iter.hasNext()) {
                SoundLayerListener listener = (SoundLayerListener) iter.next();
                listener.soundObjectRemoved(this, sObj);
            }
        }
    }

    /* CLEANUP */

    public void clearListeners() {
        if (propListeners != null) {
            propListeners.clear();
            propListeners = null;
        }

        if (layerListeners != null) {
            layerListeners.clear();
            layerListeners = null;
        }
    }

    /**
     * Generates notes for the SoundLayer, skipping over soundObjects which do
     * not contribute notes between the startTime and endTime arguments.
     * 
     * StartTime and endTime is adjusted by the PolyObject before passing into
     * SoundLayer if possible, if not possible, will adjust to render everything
     * and filter on top layer.
     */
    public NoteList generateForCSD(CompileData compileData, float startTime, 
            float endTime) throws SoundLayerException {
        
        NoteList notes = new NoteList();
        
        Collections.sort(soundObjects, sObjComparator);


        for (SoundObject sObj : soundObjects) {
            try {
            
                float sObjStart = sObj.getStartTime();
                float sObjDur = sObj.getSubjectiveDuration();
                float sObjEnd = sObjStart + sObjDur;

                if (sObjEnd > startTime) {
                    if (endTime <= startTime) {

                        float adjustedStart = startTime - sObjStart;
                        if (adjustedStart < 0.0f) {
                            adjustedStart = 0.0f;
                        }

                        notes.merge((sObj).generateForCSD(compileData, adjustedStart, -1.0f));
                    } else if (sObjStart < endTime) {

                        float adjustedStart = startTime - sObjStart;
                        float adjustedEnd = endTime - sObjStart;

                        if (adjustedStart < 0.0f) {
                            adjustedStart = 0.0f;
                        }

                        if (adjustedEnd >= sObjDur) {
                            adjustedEnd = -1.0f;
                        }

                        notes.merge((sObj).generateForCSD(compileData, adjustedStart,
                                adjustedEnd));
                    }
                }
            } catch (Exception e) {
                throw new SoundLayerException(this, "Error in SoundLayer: "
                        + this.getName(), e);
            }
        }

        try {
            ScoreUtilities.applyNoteProcessorChain(notes, this.npc);
        } catch (NoteProcessorException e) {
            throw new SoundLayerException(this, "Error in SoundLayer: "
                    + this.getName(), e);
        }
        

        return notes;
    }
}

File: packages/blue-data/src/sound-objects/sound-layer.ts
/**
 * SoundLayer — a layer within a PolyObject that contains SoundObjects.
 * Mirrors the Java SoundLayer class.
 *
 * generateForCSD sorts sound objects by start time, computes adjusted
 * start/end times relative to each object's timeline, and merges notes
 * without applying an additional offset (sound objects handle their own
 * start time internally).
 */
import { Layer, LAYER_HEIGHT } from '../score/layers/layer';
import { AutomatableLayer } from '../score/layers/automatable-layer';
import { ParameterIdList } from '../automation/parameter-id-list';
import { ScoreObject } from '../score/score-object';
import { SoundObject } from '../sound-objects/sound-object';
import { TimeContext } from '../time/time-context';
import { CompileData } from '../compile-data';
import { NoteList } from './note-list';
import { NoteProcessorChain } from '../note-processors/note-processor-chain';
import { applyNoteProcessorChain, applyNoteProcessorChainAsync } from '../utilities/score';
import { DEFAULT_LAYER_COLOR, normalizeLayerColor } from '../score/layers/layer-color';
import { Element } from '../serialization/xml-reader';

export class SoundLayer extends Array<SoundObject> implements Layer, AutomatableLayer {
  private _name = '';
  private _muted = false;
  private _solo = false;
  private _heightIndex = 0;
  private _backgroundColor = DEFAULT_LAYER_COLOR;
  private _npc = new NoteProcessorChain();
  private _automationParameters = new ParameterIdList();
  private _unknownAttributes = new Map<string, string>();
  private _unknownChildren: Element[] = [];

  constructor(other?: SoundLayer | number) {
    if (typeof other === 'number') {
      super(other);
      return;
    }

    super();
    if (other) {
      this._name = other._name;
      this._muted = other._muted;
      this._solo = other._solo;
      this._heightIndex = other._heightIndex;
      this._backgroundColor = other._backgroundColor;
      this._npc = new NoteProcessorChain(other._npc);
      this._automationParameters = other._automationParameters.deepCopy();

      for (const [k, v] of other._unknownAttributes) {
        this._unknownAttributes.set(k, v);
      }
      this._unknownChildren = other._unknownChildren.map((c) => c.clone());

      for (const sObj of other) {
        this.push(sObj.deepCopy());
      }
    }
  }

  getUnknownAttributes(): ReadonlyMap<string, string> {
    return this._unknownAttributes;
  }

  setUnknownAttribute(name: string, value: string): void {
    this._unknownAttributes.set(name, value);
  }

  getUnknownChildren(): readonly Element[] {
    return this._unknownChildren;
  }

  addUnknownChild(child: Element): void {
    this._unknownChildren.push(child);
  }

  // ─── Layer implementation ───

  getBackgroundColor(): number {
    return this._backgroundColor;
  }

  setBackgroundColor(color: number): void {
    this._backgroundColor = normalizeLayerColor(color);
  }

  getName(): string {
    return this._name;
  }

  setName(name: string): void {
    this._name = name;
  }

  getLayerHeight(): number {
    return LAYER_HEIGHT * (this._heightIndex + 1);
  }

  isMuted(): boolean {
    return this._muted;
  }

  setMuted(muted: boolean): void {
    this._muted = muted;
  }

  isSolo(): boolean {
    return this._solo;
  }

  setSolo(solo: boolean): void {
    this._solo = solo;
  }

  getHeightIndex(): number {
    return this._heightIndex;
  }

  setHeightIndex(heightIndex: number): void {
    this._heightIndex = heightIndex;
  }

  getNoteProcessorChain(): NoteProcessorChain {
    return this._npc;
  }

  setNoteProcessorChain(chain: NoteProcessorChain): void {
    this._npc = chain;
  }

  getAutomationParameters(): ParameterIdList {
    return this._automationParameters;
  }

  accepts(object: ScoreObject): boolean {
    // SoundLayer accepts any SoundObject
    return 'generateForCSD' in object;
  }

  contains(object: ScoreObject): boolean {
    return (this as SoundObject[]).includes(object as SoundObject);
  }

  remove(object: ScoreObject): boolean {
    const idx = this.indexOf(object as SoundObject);
    if (idx !== -1) {
      this.splice(idx, 1);
      return true;
    }
    return false;
  }

  clearScoreObjects(): void {
    this.length = 0;
  }

  /**
   * Generate notes for all sound objects in this layer.
   * Mirrors Java SoundLayer.generateForCSD exactly:
   * - Sort sound objects by start time
   * - Compute adjustedStart/adjustedEnd relative to each object's timeline
   * - Skip objects that don't overlap with the render range
   * - Each sound object handles its own start time offset internally
   * - Do NOT apply an additional offset here
   */
  generateForCSD(
    context: TimeContext,
    compileData: CompileData,
    startTime: number,
    endTime: number,
  ): NoteList {
    const noteList = new NoteList();

    // Sort sound objects by start time
    const sorted = [...this].sort((a, b) =>
      a.getStartTime().toBeats(context) - b.getStartTime().toBeats(context),
    );

    for (const sObj of sorted) {
      const sObjStart = sObj.getStartTime().toBeats(context);
      const sObjDur = sObj.getSubjectiveDuration().toBeats(context);
      const sObjEnd = sObjStart + sObjDur;

      // Skip objects that end before the render start
      if (sObjEnd <= startTime) continue;

      let adjustedStart: number;
      let adjustedEnd: number;

      if (endTime <= startTime) {
        // No end time constraint: render from adjustedStart to end
        adjustedStart = startTime - sObjStart;
        if (adjustedStart < 0) adjustedStart = 0;
        adjustedEnd = -1;
      } else if (sObjStart < endTime) {
        // Both start and end constraints
        adjustedStart = startTime - sObjStart;
        adjustedEnd = endTime - sObjStart;
        if (adjustedStart < 0) adjustedStart = 0;
        if (adjustedEnd >= sObjDur) adjustedEnd = -1;
      } else {
        // Object starts after render end — skip
        continue;
      }

      const nl = sObj.generateForCSD(context, compileData, adjustedStart, adjustedEnd);
      noteList.merge(nl);
    }

    return applyNoteProcessorChain(noteList, this._npc);
  }

  async generateForCSDAsync(
    context: TimeContext,
    compileData: CompileData,
    startTime: number,
    endTime: number,
  ): Promise<NoteList> {
    const noteList = new NoteList();

    const sorted = [...this].sort((a, b) =>
      a.getStartTime().toBeats(context) - b.getStartTime().toBeats(context),
    );

    for (const sObj of sorted) {
      const sObjStart = sObj.getStartTime().toBeats(context);
      const sObjDur = sObj.getSubjectiveDuration().toBeats(context);
      const sObjEnd = sObjStart + sObjDur;

      if (sObjEnd <= startTime) continue;

      let adjustedStart: number;
      let adjustedEnd: number;

      if (endTime <= startTime) {
        adjustedStart = startTime - sObjStart;
        if (adjustedStart < 0) adjustedStart = 0;
        adjustedEnd = -1;
      } else if (sObjStart < endTime) {
        adjustedStart = startTime - sObjStart;
        adjustedEnd = endTime - sObjStart;
        if (adjustedStart < 0) adjustedStart = 0;
        if (adjustedEnd >= sObjDur) adjustedEnd = -1;
      } else {
        continue;
      }

      const nl = sObj.generateForCSDAsync
        ? await sObj.generateForCSDAsync(context, compileData, adjustedStart, adjustedEnd)
        : sObj.generateForCSD(context, compileData, adjustedStart, adjustedEnd);
      noteList.merge(nl);
    }

    return applyNoteProcessorChainAsync(noteList, this._npc, compileData);
  }

  deepCopy(): SoundLayer {
    return new SoundLayer(this);
  }
}
